### PR TITLE
Linkshell Library/Service

### PR DESCRIPTION
**Pull request by Nekseus**
Minor edit to types.lua

### CODE REVIEW

File: linkshell_service/linkshell_service.lua [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+}
+  
+local linkshell = {
+    [1] = { message = { }, color = { }, },
+    [2] = { message = { }, },
+}
+
+handle_color = function(p)
+    local data = linkshell[1]
+        data.color.red = p.linkshell_red
```
Comment (z16) [inline linkshell_service/linkshell_service.lua]: The indentation is off.

File: linkshell_service/linkshell_service.lua [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+packets.incoming[0x00D]:register(handle_color)
+packets.incoming[0x037]:register(handle_color)
+packets.incoming[0x0C9]:register(handle_color)
+
+local last_00D = packets.incoming[0x00D].last
+local last_037 = packets.incoming[0x037].last
+local last_0C9 = packets.incoming[0x0C9].last
+
+local handler_color(last_00D)
```
Comment (z16) [inline linkshell_service/linkshell_service.lua]: Does this even load? The `local` here is incorrect.

Also, you should check if the packets aren't `nil` first, because they might be and the handler can't deal with that.

File: packet_service/types.lua [hunk 3]
```diff
@@ -816,9 +816,9 @@ types.incoming[0x037] = struct({
     movement_speed_half = {0x28, bit(uint16, 12), offset=0},
     yalms_per_step      = {0x2A, bit(uint16, 9), offset=0}, -- Determines how quickly your animation walks
     state               = {0x2C, state},
-    linkshell1_red      = {0x2D, uint8},
-    linkshell1_green    = {0x2E, uint8},
-    linkshell1_blue     = {0x2F, uint8},
+    linkshell_red      = {0x2D, uint8},
+    linkshell_green    = {0x2E, uint8},
+    linkshell_blue     = {0x2F, uint8},
```
Comment (z16) [inline packet_service/types.lua]: Why did you remove the `1` from here? Isn't it only for the first linkshell? (And if it is, adjust the spacing so the `=` line up.)

Comment (Nekseus) [reply packet_service/types.lua]: because the other 2 packets with the same data are written w/o a 1. it's either this or one and the service handles it this way since its 2 to 1.

**Comment by Nekseus** [comment]
@z16 senpai notice me!

File: linkshell_service/linkshell_service.lua [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+packets.incoming[0x0E0]:register(function(p)
+    local data = linkshell[p.linkshell_number]
+    data.bag_index = p.bag_index
+end)
+
+packets.incoming[0x00D]:register(handle_color)
+packets.incoming[0x037]:register(handle_color)
+packets.incoming[0x0C9]:register(handle_color)
+
+handler_color(packets.incoming[0x00D].last)
```
Comment (z16) [inline linkshell_service/linkshell_service.lua]: Check if `.last` is `nil` before calling the handler. See what the `account_service` does here.

File: linkshell_service/linkshell_service.lua [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    next = next,
+}
+  
+local linkshell.data = {
+    [1] = { message = { }, color = { }, },
+    [2] = { message = { }, },
+}
+
+local handle_color = function(p)
+local data = linkshell.data[1]
```
Comment (z16) [inline linkshell_service/linkshell_service.lua]: Indent this line.

File: linkshell_service/linkshell_service.lua [hunk 6]
```diff
@@ -0,0 +1,72 @@
+local packets = require('packets')
+local shared = require('shared')
+
+local linkshell = shared.new('linkshell')
+
+-- Does this file need this?
```
Comment (z16) [inline linkshell_service/linkshell_service.lua]: It currently needs it, though it's debatable whether it *should* need it. Though I'd say that's for another time, so the comment is no longer necessary.

File: packet_service/types.lua (hunk 3 above)
Comment (z16) [reply packet_service/types.lua]: I would leave it as `linkshell1` and adjust it in other packets to be more correct. Some of these definitions were written before LS2 was even a thing, so it could just be an oversight.

File: linkshell_service/linkshell_service.lua (hunk 6 above)
Comment (Nekseus) [reply linkshell_service/linkshell_service.lua]: https://github.com/Windower/packages/pull/63/commits/7d40d96e9dc89d3258969f887883bf12cf58c16a

**Comment by Nekseus** [comment]
https://github.com/Windower/packages/pull/63/files

File: linkshell_service/linkshell_service.lua [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+    next = next,
+}
+  
+local linkshell.data = {
+    [1] = { lsmes = { }, color = { }, },
+    [2] = { lsmes = { }, },
+}
+
+local handler_color = function(p)
+local data = linkshell[1]
```
Comment (z16) [inline linkshell_service/linkshell_service.lua]: This line needs to be indented.

File: linkshell_service/linkshell_service.lua [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+end)
+
+packets.incoming[0x0E0]:register(function(p)
+    local data = linkshell[p.linkshell_number]
+    data.bag_index = p.bag_index
+end)
+
+packets.incoming[0x00D]:register(handler_color)
+packets.incoming[0x037]:register(handler_color)
+packets.incoming[0x0C9][0x01]:register(handler_color) -?? types.lua:1796/1806
```
Comment (z16) [inline linkshell_service/linkshell_service.lua]: I assume that was meant to be a comment? As it is now it doesn't load.

File: linkshell_service/linkshell_service.lua [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+
+handler_color(packets.incoming[0x00D].last)
```
Comment (z16) [inline linkshell_service/linkshell_service.lua]: You need to check these for `nil`. See how `account_service` does it
```lua
local last_00D = packets.incoming[0x00D].last
local last_037 = packets.incoming[0x037].last
local last_0C9_01 = packets.incoming[0x0C9][0x01].last

if last_00D then
    handler_color(last_00D)
end
if last_037 then
    handler_color(last_037)
end
if last_0C9_01 then
    handler_color(last_0C9_01)
end
```

File: linkshell_service/linkshell_service.lua [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+
+last_00D = packets.incoming[0x00D].last
+last_037 = packets.incoming[0x037].last
+
+if last_00D then
+    handle_color(last_00D)
+end
+if last_037 then
+    handle_color(last_00B)
+end
```
Comment (z16) [inline linkshell_service/linkshell_service.lua]: I cannot imagine the `0x00D` packet being useful here. All LS info changes should be in `0x0CC` and `0x0E0` imo, while `0x037` also has LS info in it, I don't think it can ever be different or more useful than `0x0CC`. I would remove everything but those two, and adjust `last_XXX` accordingly. Also, make the last packets `local`.

This has the advantage that we also don't need to worry about the order, since they carry entirely separate information. I would say go this route until we notice errors with this approach, which I don't think there should be.

File: packet_service/types.lua [hunk 11]
```diff
@@ -1866,9 +1866,9 @@ types.incoming[0x0C9] = multiple({
     [0x01] = struct({
         icon_set_subtype= {0x0A, uint8},
         icon_set_id     = {0x0B, uint8},
-        linkshell_red   = {0x0C, bit(uint16, 4), offset=0},
-        linkshell_green = {0x0C, bit(uint16, 4), offset=4},
-        linkshell_blue  = {0x0C, bit(uint16, 4), offset=8},
+        linkshell1_red   = {0x0C, bit(uint16, 4), offset=0},
```
Comment (z16) [inline packet_service/types.lua]: Align the assignments again.

File: linkshell_service/linkshell_service.lua [hunk 12]
```diff
@@ -0,0 +1,66 @@
+local packets = require('packets')
+local shared = require('shared')
+local linkshell = shared.new('linkshell')
+
+linkshell.env = {
+    next = next,
+}
+
```
Comment (Nifim) [inline linkshell_service/linkshell_service.lua]: There is a space here.

File: linkshell/linkshell.lua [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+        if type(result) ~= 'table' then
+            return result
+        end
+
+        return setmetatable({}, {
+            __index = function(_, l)
+                __index = function(_, l)
+                    return result[l]
+                end,
+            __newindex = function() error('This value is read-only.') end,
```
Comment (Nifim) [inline linkshell/linkshell.lua]: indenting here is off

Comment (Nifim) [reply linkshell/linkshell.lua]: :/ is this just indenting off? something seems weird with this meta table.

File: linkshell/linkshell.lua [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+    __index = function(_, k)
+        local result = fetch_data:read(k)
+
+        if type(result) ~= 'table' then
+            return result
+        end
+
+        return setmetatable({}, {
+            __index = function(_, l)
+                __index = function(_, l)
```
Comment (cairface) [inline linkshell/linkshell.lua]: Why is __index defined within index o.o
